### ISSUE TITLE
Thumbnails scale of 1 to 1

### ISSUE THREAD

**Issue by faq4272** [lrsjng/h5ai#646]
Hello!
How to set class-thumb.php so that thumbnails are created in a scale of 1 to 1, rather than cropped to a 4x3 format. Even if they still will be within the 4x3 window, but just to display the entire contents of the image on the right scale. It is extremely inconvenient that some of the high images are hidden somewhere below.

**Comment by Armstrong1992** [comment]
So you want to see whole image in thumbnail, not just a fragment? And you realize that for bigger images it will result in really weird image. 

I could dig into the code to see what to change. When i find some some time :)

**Comment by faq4272** [comment]
@Armstrong1992 Yes, I understand, but I have a lot of extended panoramic photos, or just very high ones, and it would be very convenient for me to see them in full and not to open each separately. I myself found where to change the scale and with the help of it I made an analog of viewing instagram. But since the images are cut around the edges, so far from this little.

**Comment by Armstrong1992** [comment]
Change:
https://github.com/lrsjng/h5ai/blob/44e53ab213d37f45df2acac01e957c67ec06f8a7/src/_h5ai/private/php/ext/class-thumb.php#L217
To:
```php
imagecopyresampled($this->dest, $this->source, 0, 0, 0, 0, $width, $height, $this->width, $this->height);

**Comment by faq4272** [comment]
@Armstrong1992 Thank you, it's better now, but is it possible to preserve the aspect ratio? since now it turns out that 16: 9 images are compressed into a 4: 3 frame. Is it possible to fully see the images in the same 4: 3 frame but with the preservation of proportions, and leave the empty space empty or fill it with some kind of color. Thanks for the help.

**Comment by Armstrong1992** [comment]
```php
        if ($this->width > $this->height) {
            $thumb_w = $width;
            $thumb_h = $this->height * ($height / $this->width);
        } elseif ($this->width < $this->height) {
            $thumb_w = $this->width * ($width / $this->height);
            $thumb_h = $height;
        } elseif ($this->width == $this->height) {
            $thumb_w = $width;
            $thumb_h = $height;
        }

        $this->dest = imagecreatetruecolor($width, $height);
        $icol       = imagecolorallocate($this->dest, 255, 255, 255);
        imagefill($this->dest, 0, 0, $icol);
        imagecopyresampled($this->dest, $this->source, 0, 0, 0, 0, $thumb_w, $thumb_h, $this->width, $this->height);

**Comment by faq4272** [comment]
Cool, thanks a lot. I corrected it a little, because the images were still slightly flattened.

        if ($this->width > $this->height) {
            $thumb_w = $width;
            $thumb_h = ($this->height * ($height / $this->width)) * 1.25;
        } elseif ($this->width < $this->height) {
            $thumb_w = ($this->width * ($width / $this->height)) * 0.75;
            $thumb_h = $height;
        } elseif ($this->width == $this->height) {
            $thumb_w = $width * 0.75;
            $thumb_h = $height;
        }

        $this->dest = imagecreatetruecolor($width, $height);
        $icol       = imagecolorallocate($this->dest, 255, 255, 255);
        imagefill($this->dest, 0, 0, $icol);
        imagecopyresampled($this->dest, $this->source, 0, 0, 0, 0, $thumb_w, $thumb_h, $this->width, $this->height);

As though now still to fix quality of the image that it was possible to disassemble we will tell the text on a preview without opening the image, for example for comics, or simply very detailed images. Quality like everywhere put in 100, but I see that the image still very much shrinks. can I set something like a thumbnail resolution?
